File: impls/cc/step6_file.cc
#include "core.hh"
#include "env.hh"
#include "printer.hh"
#include "reader.hh"
#include <iostream>
#include <map>
#include <string>

std::shared_ptr<MalType> eval(std::shared_ptr<MalType> input, std::shared_ptr<Env> env);

std::shared_ptr<MalType> read(const std::string &input)
{
    return read_str(input);
}

std::shared_ptr<MalType> eval_ast(std::shared_ptr<MalType> ast, std::shared_ptr<Env> env)
{
    switch (ast->type())
    {
    case MalType::Type::Symbol:
    {
        auto symbol = static_cast<MalSymbol &>(*ast);
        if (symbol.is_string() || symbol.is_keyword())
            return ast;

        return env->get(symbol);
    }
    case MalType::Type::List:
    {
        auto &list = static_cast<MalList &>(*ast);
        auto new_list = std::make_shared<MalList>(list.lparen(), list.rparen());
        for (auto &a : list)
        {
            auto l = eval(a, env);
            if (!l)
                return nullptr;
            new_list->push_back(l);
        }
        return new_list;
    }
    default:
        return ast;
    }
}

std::shared_ptr<MalType> eval(std::shared_ptr<MalType> input, std::shared_ptr<Env> env)
{
    while (true)
    {
        if (!input)
            return nullptr;

        if (input->type() != MalType::Type::List)
            return eval_ast(input, env);

        auto &list = static_cast<MalList &>(*input);
        if (list.empty())
            return input;

        if (!list.is_list())
            return eval_ast(input, env);

        if (list[0]->type() == MalType::Type::Symbol)
        {
            std::string symbol = static_cast<const MalSymbol &>(*list[0]);
            if (symbol == "def!")
            {
                std::string key = static_cast<const MalSymbol &>(*list[1]);
                auto val = eval(list[2], env);
                if (!val)
                    return nullptr;
                env->set(key, val);
                return val;
            }

            if (symbol == "let*")
            {
                auto new_env = std::make_shared<Env>(env);
                auto &bindings = static_cast<const MalList &>(*list[1]);
                for (unsigned i = 0; i < bindings.size(); i += 2)
                {
                    std::string key = static_cast<const MalSymbol &>(*bindings[i]);
                    auto val = eval(bindings[i + 1], new_env);
                    if (!val)
                        return nullptr;
                    new_env->set(key, val);
                }
                input = list[2];
                env = new_env;
                continue;
            }

            if (symbol == "do")
            {
                for (unsigned i = 1; i < list.size() - 1; ++i)
                    eval(list[i], env);
                input = list[list.size() - 1];
                continue;
            }

            if (symbol == "if")
            {
                auto cond = eval(list[1], env);
                if (cond && cond->type() == MalType::Type::Symbol)
                {
                    auto condition = static_cast<MalSymbol &>(*cond);
                    if (!condition)
                    {
                        if (list.size() > 3)
                        {
                            input = list[3];
                            continue;
                        }
                        else
                            return env->get("nil");
                    }
                }
                input = list[2];
                continue;
            }

            if (symbol == "fn*")
            {
                auto closure = [list, env](std::vector<std::shared_ptr<MalType>> params)
                {
                    auto new_env = std::make_shared<Env>(static_cast<const MalList &>(*list[1]), params, env);
                    return eval(list[2], new_env);
                };
                return std::make_shared<MalFunc>(closure, list[2], list[1], env);
            }
        }

        auto plist = eval_ast(input, env);
        if (!plist)
            return nullptr;

        auto &new_list = static_cast<MalList &>(*plist);
        auto &func = static_cast<const MalFunc &>(*new_list[0]);

        std::vector<std::shared_ptr<MalType>> args;
        for (unsigned i = 1; i < new_list.size(); ++i)
            args.push_back(new_list[i]);

        if (func.is_fn())
        {
            input = func.ast();
            env = std::make_shared<Env>(static_cast<const MalList &>(*func.params()), args, func.env());
            continue;
        }

        return func(args);
    }
}

std::string print(std::shared_ptr<MalType> input)
{
    return pr_str(input);
}

std::string rep(const std::string &input, std::shared_ptr<Env> env)
{
    auto read_result = read(input);
    auto eval_result = eval(read_result, env);
    auto print_result = print(eval_result);
    return print_result;
}

int main(int argc, char *argv[])
{
    auto repl_env = std::make_shared<Env>(nullptr);

    for (auto &[key, value] : ns())
        repl_env->set(key, value);

    auto closure = [repl_env](std::vector<std::shared_ptr<MalType>> ast)
    {
        return eval(ast[0], repl_env);
    };
    repl_env->set("eval", std::make_shared<MalFunc>(closure));

    rep("(def! not (fn* (a) (if a false true)))", repl_env);
    rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\")))))", repl_env);

    auto args = std::make_shared<MalList>('(', ')');
    for (int i = 2; i < argc; ++i)
        args->push_back(read('"' + std::string(argv[i]) + '"'));
    repl_env->set("*ARGV*", args);

    if (argc > 1)
    {
        rep("(load-file \"" + std::string(argv[1]) + "\")", repl_env);
        return 0;
    }

    while (!std::cin.eof())
    {
        std::string input;
        std::cout << "user> ";
        std::getline(std::cin, input);
        auto rep_result = rep(input, repl_env);
        std::cout << rep_result << std::endl;
    }

    return 0;
}

File: impls/cs.2/step6_file.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace mal
{
    class step6_file
    {

        static MalType READ(string input)
        {
            return Reader.read_str(input);
        }

        public static MalType EVAL(MalType ast, Env env)
        {
            while (true)
            {
                if (ast is not MalList)
                {
                    return eval_ast(ast, env);
                }
                else
                {
                    MalList astList = (MalList)ast;

                    // Remove any comments from the ast
                    List<MalType> nonComments = astList.items.Where(it => !(it is MalSymbol && ((MalSymbol)it).value.StartsWith(";"))).ToList();
                    astList = new MalList(nonComments);

                    if (astList.items.Count == 0)
                    {
                        return astList;
                    }
                    else
                    {

                        MalType first = astList.items.First();
                        if (first is MalSymbol)
                        {
                            MalSymbol firstSymbol = (MalSymbol)first;
                            if (firstSymbol.value == "def!")
                            {
                                MalSymbol symbol = (MalSymbol)astList.items[1];
                                MalType value = EVAL(astList.items[2], env);
                                env.set(symbol, value);
                                return value;
                            }
                            else if (firstSymbol.value == "let*")
                            {
                                MalSeq bindings = (MalSeq)astList.items[1];
                                MalType expression = astList.items[2];
                                Env newEnv = new Env(env);
                                for (int i = 0; i < bindings.items.Count; i += 2)
                                {
                                    MalSymbol name = (MalSymbol)bindings.items[i];
                                    MalType value = EVAL(bindings.items[i + 1], newEnv);
                                    newEnv.data.Add(name, value);
                                }
                                env = newEnv;
                                ast = expression;
                                continue;
                            }
                            else if (firstSymbol.value == "do")
                            {
                                int butLast = astList.items.Count - 2;

                                // produce side-effects and then just forget
                                List<MalType> sideEffects = astList.items.Skip(1).Take(butLast).ToList();
                                eval_ast(new MalList(sideEffects), env);

                                ast = astList.items.Last();
                                continue;
                            }
                            else if (firstSymbol.value == "if")
                            {
                                MalType test = astList.items[1];
                                MalType testResult = EVAL(test, env);
                                if (testResult == MalNil.MAL_NIL || testResult == MalBoolean.MAL_FALSE)
                                {
                                    ast = (astList.items.Count > 3) ? astList.items[3] : MalNil.MAL_NIL;
                                }
                                else
                                {
                                    ast = astList.items[2];
                                }
                                continue;
                            }
                            else if (firstSymbol.value == "fn*")
                            {
                                MalSeq argNames = (MalSeq)astList.items[1];
                                MalType funcBody = astList.items[2];
                                List<MalSymbol> argSymbs = new List<MalSymbol>();
                                foreach (MalType arg in argNames.items) { if (arg is MalSymbol) argSymbs.Add((MalSymbol)arg); }
                                MalFunction fn = new MalFunction(
                                    (IList<MalType> argValues) =>
                                    {
                                        Env funcEnv = new Env(env, argSymbs, argValues);
                                        return EVAL(funcBody, funcEnv);
                                    }
                                );
                                return new MalFnTco(funcBody, argSymbs, env, fn);
                            }

                        }

                        MalType evaluated = eval_ast(ast, env);
                        if (evaluated is not MalList) return evaluated;
                        MalList evaluatedList = (MalList)evaluated;

                        // Function application
                        MalType funcFirst = evaluatedList.items[0];
                        if (funcFirst is MalFunction)
                        {
                            MalFunction func = (MalFunction)funcFirst;
                            MalType retVal = func.function(evaluatedList.items.Skip(1).ToList());
                            return retVal;
                        }
                        else // MalFnTco
                        {
                            MalFnTco fnTco = (MalFnTco)funcFirst;
                            List<MalType> fnArgs = evaluatedList.items.Skip(1).ToList();
                            Env newEnv = new Env(fnTco.env, fnTco.@params, fnArgs);
                            ast = fnTco.ast;
                            env = newEnv;
                            continue;
                        }
                    }
                }
            }
        }

        static string PRINT(MalType input)
        {
            return printer.pr_str(input, true);
        }

        static string rep(string input)
        {
            MalType read = READ(input);
            MalType evaled = EVAL(read, repl_env);
            string printed = PRINT(evaled);
            return printed;
        }

        public static MalType eval_ast(MalType ast, Env env)
        {
            if (ast is MalSymbol)
            {
                MalSymbol astSymbol = (MalSymbol)ast;
                return env.get(astSymbol);
            }
            else if (ast is MalList)
            {
                MalList astList = (MalList)ast;
                List<MalType> evaluated = astList.items.Select(item => EVAL(item, env)).ToList();
                return new MalList(evaluated);
            }
            else if (ast is MalVector)
            {
                MalVector astVector = (MalVector)ast;
                List<MalType> evaluated = astVector.items.Select(item => EVAL(item, env)).ToList();
                return new MalVector(evaluated);
            }
            else if (ast is MalHashmap)
            {
                MalHashmap astHashmap = (MalHashmap)ast;
                Dictionary<MalType, MalType> newKVs = new Dictionary<MalType, MalType>();
                foreach (KeyValuePair<MalType, MalType> kv in astHashmap.values)
                {
                    newKVs.Add(EVAL(kv.Key, env), EVAL(kv.Value, env));
                }
                return new MalHashmap(newKVs);
            }
            else
            {
                return ast;
            }
        }

        public static Env repl_env = new Env(null);

        static void Main(string[] args)
        {
            // Load the built-in functions
            foreach (var pair in core.ns) { repl_env.set(new MalSymbol(pair.Key), pair.Value); }

            // Define 'eval'
            repl_env.data.Add(new MalSymbol("eval"), new MalFunction((IList<MalType> args) =>
            {
                MalType ast = args[0];
                return EVAL(ast, repl_env);
            }));

            // Functions defined on Mal itself:
            // - define 'not'
            rep("(def! not (fn* (a) (if a false true)))");
            // - define 'load-file'
            rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\"))))))");

            // ARGV
            List<MalType> malArgs = new List<MalType>();
            malArgs.AddRange(args.Skip(1).Select(arg => new MalString(arg)).ToList());
            repl_env.data.Add(new MalSymbol("*ARGV*"), new MalList(malArgs));

            // if called with arguments, treat first as a script name
            if (args.Length > 0)
            {
                rep("(load-file \"" + args[0] + "\")");
                return;
            }

            // TESTS
            // rep("(load-file \"../../tests/computations.mal\")");

            string line = null;
            do
            {
                Console.Write("user> ");
                line = Console.ReadLine();
                if (line != null)
                {
                    try
                    {
                        Console.WriteLine(rep(line));
                    }
                    catch (Exception ex)
                    {

                        Console.WriteLine(ex.Message);
                    }
                }
            } while (line != null);
            Console.WriteLine();
        }
    }
}
